README.md

```diff
@@ -0,0 +1,39 @@
+# Initrd for Ubuntu Core
+
+This repository contains source code to build initial ramdisk used in Ubuntu Core Embedded Linux OS.
+
+# Purpose
+
+See: https://en.wikipedia.org/wiki/Initial_ramdisk
+
+# Architecture
+
+In Ubuntu Core, initrd.img is part of Kernel Snap. See [integration](#integrating-with-kernel-snap) below.
+
+In UC20, initrd is migrated from script based implementation (UC16/18) to systemd based.
+
+# Building
+
+# Testing & Debugging
+
+In order to test your unit files in the initrd shell, you need to add
+
+```
+rd.systemd
```

Comment:
Most systemd initrd behaviour can be changed using kernel command line options. Unlike the running system, all the kernel command line options start with "rd.systemd" prefix, instead of the usual "systemd" one.

Examples:

* `rd.systemd.unit=emergency.target dangerous` Boot into emergency shell. To get out of it, one will need to start default.target.
* `rd.systemd.debug-shell=true dangerous` enable debug shell on tty9.

Reply:
First I tried on my RPi-4B:
```
wc_otg.lpm_enable=0 console=serial0,115200 elevator=deadline rng_core.default_quality=700 vt.handoff=2 quiet splash dangerous rd.systemd.unit=emergency.shell rd.systemd.debug-shell=true rd.systemd.unit=emergency.target
```

Save & Sync and it does not enter any shell, just boots into console-conf as follows:

```
Decompressing kernel...
Uncompressed size: 23843328 = 0x16BD200
19997970 bytes read in 1461 ms (13.1 MiB/s)
Booting Ubuntu (with booti) from mmc 0:...
## Flattened Device Tree blob at 02600000
   Booting using the fdt blob at 0x2600000
   Using Device Tree in place at 0000000002600000, end 000000000260f077

Starting kernel ...

[    1.216004] spi-bcm2835 fe204000.spi: could not get clk: -517
[    2.973773] the-tool[218]: 2021/06/07 07:30:52.553398 cmdline.go:80: ignoring recovery system label "
20210607" in "run" mode
[    4.966172] the-tool[218]: 2021/06/07 07:30:54.545802 secboot_tpm.go:222: cannot open TPM connection:
 no TPM2 device is available
[   10.520255] brcmfmac: brcmf_fw_alloc_request: using brcm/brcmfmac43455-sdio for chip BCM4345/6
[   11.055302] brcmfmac: brcmf_fw_alloc_request: using brcm/brcmfmac43455-sdio for chip BCM4345/6
[   11.160024] brcmfmac: brcmf_c_preinit_dcmds: Firmware: BCM4345/6 wl0: Jan  4 2021 19:56:29 version 7.
45.229 (617f1f5 CY) FWID 01-2dbd9d2e


[   17.586846] brcmfmac: brcmf_cfg80211_set_power_mgmt: power save enabled
Ubuntu Core 20 on 192.168.1.224 (ttyS0)
The host key fingerprints are:

    RSA     SHA256:XXXXXXXXXXXXXXXXXX
    ECDSA   SHA256:tXXXXXXXXXXXXXX
    ED25519 SHA256:yXXXXXXXXXXXXXX

To login:

    ssh stulluk@192.168.1.224

Personalize your account at https://login.ubuntu.com.

```

And then, I tried this:
```
wc_otg.lpm_enable=0 console=serial0,115200 elevator=deadline rng_core.default_quality=700 vt.handoff=2 quiet splash systemd.debug_shell=serial0 console=tty0 dangerous
```
This allows a command prompt in initrd to appear on my serial console, but then, if I understand correctly, it pivot-roots and jumps to systemd in real rootfs and starts console-conf:

```
Decompressing kernel...
Uncompressed size: 23843328 = 0x16BD200
19997970 bytes read in 1461 ms (13.1 MiB/s)
Booting Ubuntu (with booti) from mmc 0:...
## Flattened Device Tree blob at 02600000
   Booting using the fdt blob at 0x2600000
   Using Device Tree in place at 0000000002600000, end 000000000260f043

Starting kernel ...

[    1.211564] spi-bcm2835 fe204000.spi: could not get clk: -517
root@ubuntu:/# [   10.527158] brcmfmac: brcmf_fw_alloc_request: using brcm/brcmfmac43455-sdio for chip B
CM4345/6
[   10.887789] brcmfmac: brcmf_fw_alloc_request: using brcm/brcmfmac43455-sdio for chip BCM4345/6
[   10.945672] brcmfmac: brcmf_c_preinit_dcmds: Firmware: BCM4345/6 wl0: Jan  4 2021 19:56:29 version 7.
45.229 (617f1f5 CY) FWID 01-2dbd9d2e
root@ubuntu:/#



[   17.709294] brcmfmac: brcmf_cfg80211_set_power_mgmt: power save enabled
Ubuntu Core 20 on 192.168.1.224 (ttyS0)
The host key fingerprints are:

    RSA     SHA256:XXXXXXXXXXXXXXXXXX
    ECDSA   SHA256:tXXXXXXXXXXXXXX
    ED25519 SHA256:yXXXXXXXXXXXXXX

To login:

    ssh stulluk@192.168.1.224

Personalize your account at https://login.ubuntu.com.


Ubuntu Core 20 on 192.168.1.224 (ttyS0)
The host key fingerprints are:

 RSA     SHA256:XXXXXXXXXXXXXXXXXX
    ECDSA   SHA256:tXXXXXXXXXXXXXX
    ED25519 SHA256:yXXXXXXXXXXXXXX

To login:

    ssh stulluk@192.168.1.224

Personalize your account at https://login.ubuntu.com.
```

Finally, to avoid starting console-conf, I modified cmdline.txt as follows:
```
dwc_otg.lpm_enable=0 console=serial0,115200 elevator=deadline rng_core.default_quality=700 vt.handoff=2 quiet splash systemd.debug_shell=serial0 console=tty0 dangerous systemd.mask=serial-getty@.service systemd.mask=serial-console-conf@.service systemd.mask=console-conf@.service
```
Save & sync & reboot:

```
Decompressing kernel...
Uncompressed size: 23843328 = 0x16BD200
19997970 bytes read in 1462 ms (13 MiB/s)
Booting Ubuntu (with booti) from mmc 0:...
## Flattened Device Tree blob at 02600000
   Booting using the fdt blob at 0x2600000
   Using Device Tree in place at 0000000002600000, end 000000000260f0b3

Starting kernel ...

[    1.211515] spi-bcm2835 fe204000.spi: could not get clk: -517
root@ubuntu:/# [   10.467993] brcmfmac: brcmf_fw_alloc_request: using brcm/brcmfmac43455-sdio for chip B
CM4345/6
[   10.813163] brcmfmac: brcmf_fw_alloc_request: using brcm/brcmfmac43455-sdio for chip BCM4345/6
[   10.840847] brcmfmac: brcmf_c_preinit_dcmds: Firmware: BCM4345/6 wl0: Jan  4 2021 19:56:29 version 7.
45.229 (617f1f5 CY) FWID 01-2dbd9d2e

root@ubuntu:/# 
root@ubuntu:/# 
root@ubuntu:/# 
```

And it is quite strange that, I have been able to ssh to my pi -> It means I am still not in the initrd, it just pivot-root to main..

Could you please guide me, what I am missing here?

Reply:
It seems somehow cmdline gets overwritten:
```
stulluk@ubuntu:~$ cat /proc/cmdline 
 coherent_pool=1M 8250.nr_uarts=1 snd_bcm2835.enable_compat_alsa=0 snd_bcm2835.enable_hdmi=1 video=HDMI-A-1:1920x1080M@60,margin_left=48,margin_right=48,margin_top=48,margin_bottom=48 smsc95xx.macaddr=E4:5F:XX:XX:CB:2A vc_mem.mem_base=0x3ec00000 vc_mem.mem_size=0x40000000  dwc_otg.lpm_enable=0 console=ttyS0,115200 elevator=deadline rng_core.default_quality=700 vt.handoff=2 quiet splash systemd.debug_shell=ttyS0 console=tty0 dangerous rd.systemd.unit=emergency.shell quiet splash snapd_recovery_mode=run snapd_recovery_system=20210607 panic=-1 systemd.gpt_auto=0 rd.systemd.unit=basic.target
```
Who adds the last "rd.systemd.unit=basic.target" ?

Reply:
overall the cmdline you show looks very old.... Can you please elaborate which image are you using, which revision numbers of all the snaps involved? I.e. the full build info file, if any. or like the list of the names of all the snaps from the seed partition?

Reply:
Found the reason as I mentioned above.
Problem is boot.scr
boot.scr appends "rd.systemd.unit=basic.target" which causes initrd to jump / pivot-root
editing boot.scr fixed it for me. I have been able to enter busybox shell.
I will create a seperate document to explain how to do it.

Reply:
BTW, this is the latest edge image 3 days ago, downloaded from cdimage :)

```
stulluk@ubuntu:~$ snap list
Name       Version               Rev    Tracking     Publisher   Notes
core20     20210611              1054   latest/edge  canonical✓  base
pi         20-1                  97     20/edge      canonical✓  gadget
pi-kernel  5.4.0-1037.40         292    20/edge      canonical✓  kernel
snapd      2.51+git378.gfb31590  12381  latest/edge  canonical✓  snapd
stulluk@ubuntu:~$ 
```

Reply:
```
boot.scr appends "rd.systemd.unit=basic.target" which causes initrd to jump / pivot-root
editing boot.scr fixed it for me. I have been able to enter busybox shell.
I will create a seperate document to explain how to do it.
```

Please do not create new documents.... this is a bug, please find where boot.scr comes from and fix it such that it doesn't contain that anymore.

Reply:
boot.scr comes from the flash-kernel deb in a PPA that @waveform80 manages that I can never remember the name of :smile:

Reply:
> Please do not create new documents.... this is a bug, please find where boot.scr comes from and fix it such that it doesn't contain that anymore.

well, we have a kind of unwritten rule to have all Ubuntu Core bits have a proper README that at least explains the basics of what this source tree is about (see the gadget trees etc), how to build it and potentially also how to debug it ... 

this new code is a glaring exception, the only place where any information is currently stored are your and ians brains, @stulluk is just trying to fix an omission here ...

Reply:
@xnox also note that this is *not* a bug, we definitely do want the enforced basic.target in all production setups (not sure if this need to be enforced by a cmdline option though), like we took some effort to disable the break= option in the former initrd implementation, what @stulluk tries to describe is a way of debugging for people needing to do this during an enablement or for checking bugfixes at runtime, like we had it described for initramfs-tools before... admittedly this could go into a HACKING.md if preferred but we do need documentation around all this that has not been created yet and it would be helpful if you could constructively help with your knowledge here ...

Reply:
@ogra1 production gadgets must not specify any rd.systemd options. So something is wrong....

Reply:
https://bugs.launchpad.net/ubuntu/+source/flash-kernel/+bug/1933093